### PR TITLE
NyanCat v1.0.0

### LINKED ISSUE #39: Add a nyancat reporter
Cause every good test framework needs more cat's we should add an nyancat-reporter.

This also would showcase and harden the API surface for reporters as well as custom and user made plugins.

See [jest-nyancat-reporter](https://github.com/philihp/jest-nyancat-reporter) as an inspriation.

### PR DESCRIPTION
**Pull request by hazard999**
First draft of a NyanCatReporter

Fixes #39

### CODE REVIEW

**Comment by biohazard999** [comment]
To test out the reporter you need to register the reporter in `Tasty` and don't run the `b` command cause it launches 2 runs in parallel, cause we don't solved the register multiple reporters problem yet.

```cs

  public static class Tasty
    {
        static Tasty()
        {
            DefaultScope = new TastyScope();
            // ConsoleReporter.Register();
            NyanCatReporter.Register();
            DefaultScope.RegisterTransport(NamedPipeRemoteHook.CreateNamedPipeTransportStream);
        }
```

`C:\F\git\Tasty\test\Xenial.Tasty.Tests`

```cmd
dotnet run -f netcoreapp3.1
```

**Comment by biohazard999** [comment]
Currently the output of nyan is a little bit *deformed* :P

image

**Comment by hazard999** [comment]
> Currently the output of nyan is a little bit _deformed_ :P
> 
> image

At least it nyans 🤷‍♂️😆

File: src/Xenial.Tasty/Xenial.Tasty.csproj [hunk 1]
```diff
@@ -11,6 +11,7 @@
   <ItemGroup>
     <PackageReference Include="StreamJsonRpc" Version="2.4.48" />
     <PackageReference Include="System.Text.Encoding.CodePages" Version="4.7.1" />
+    <PackageReference Include="Pastel" Version="1.3.2" />
```
[the remark this answers is not in the document]
Comment (hazard999) [reply src/Xenial.Tasty/Xenial.Tasty.csproj:14]: We should introduce a console abstraction for the console reporters
Would also make it easier to test a reporter

File: src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+        private static Task ReportSummary(IEnumerable<TestCase> tests)
+        {
+            foreach (var test in tests)
+                Draw(test);
+
+            return Task.CompletedTask;
+        }
+
+        public static Task Report(TestCase test)
```
[the remark this answers is not in the document]
Comment (hazard999) [reply src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs:61]: For Nyan we don't need 1/n, its designed to fill the width to 75% and then parallax the rainbow

File: src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            trajectoryWidthMax = width - nyanCatWidth;
+        }
+
+        public static void Register()
+        {
+            Tasty.RegisterReporter(Report)
+                 .RegisterReporter(ReportSummary);
+        }
+
+        private static Task ReportSummary(IEnumerable<TestCase> tests)
```
[the remark this answers is not in the document]
Comment (hazard999) [reply src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs:53]: Maybe we should split the reporting as nyan doesn't have it's own summary

File: src/Xenial.Tasty/Xenial.Tasty.csproj (hunk 1 above)
[the remark this answers is not in the document]
Comment (hazard999) [reply src/Xenial.Tasty/Xenial.Tasty.csproj:14]: something like [https://github.com/goblinfactory/konsole/blob/master/readme-iconsole.md](https://github.com/goblinfactory/konsole/blob/master/readme-iconsole.md)

[the remark this answers is not in the document]
Comment (hazard999) [reply src/Xenial.Tasty/Xenial.Tasty.csproj:14]: or we use ncurses [https://github.com/MV10/dotnet-curses/](https://github.com/MV10/dotnet-curses/)

[the remark this answers is not in the document]
Comment (biohazard999) [reply src/Xenial.Tasty/Xenial.Tasty.csproj:14]: In the `Tasty.Cli` in studio mode, we need to write to a `TextView` or `LogView` that uses ncurses under the hood (`Terminal.Gui`). The abstraction ticket is #40. Altough I have no clear idea what level of abstraction we should use here. Basically we need either some kind of protocol or abstraction that supports colors and all the cursor movement over a remote.

File: src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (biohazard999) [reply src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs:53]: If you don't need a summary simply noop

File: src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (biohazard999) [reply src/Xenial.Tasty/Reporters/Reporters.NyanCat.cs:61]: So we are going like:

```txt
1st test
_-nyanface
2nd test
_-_-nyanface
bounds limit after several tests
/overflow rainbow cause too many tests for screen width/_-_-_-nyanface
```
?

File: src/Xenial.Tasty/Xenial.Tasty.csproj (hunk 1 above)
[the remark this answers is not in the document]
Comment (biohazard999) [reply src/Xenial.Tasty/Xenial.Tasty.csproj:14]: Maybe its the best to just write to a stream under the hood. But i think a more high level abstraction would be nice though

**Comment by biohazard999** [comment]
When running build locally it blows up on the integration tests with a `HandleInvalid` in the `SetCursor`.

For now we should be fine by logging out the exception to stdout in the `ReportTestMiddleware`

```cs
public static class ReportTestMiddleware
    {
        public static TestExecutor UseTestReporters(this TestExecutor executor)
            => executor.Use(async (context, next) =>
            {
                try
                {
                    await next();
                }
                finally
                {
                    try
                    {
                        await context.CurrentScope.Report(context.CurrentCase);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"Reporter failed: {ex}");
                    }
                }
            });
    }
```

**Comment by biohazard999** [comment]
Besides the lacking summary reporter this looks awesome! Thanks @hazard999 ♥

**Pull request merged**